cdc/log.cc

```diff
@@ -1299,6 +1303,35 @@ class transformer final {
                     }
                     return make_lw_shared<cql3::untyped_result_set>(*result_set);
         });
+      } catch (exceptions::unavailable_exception& e) {
+          // `query` can throw `unavailable_exception`, which triggers an error code often interpreted by
+          // clients as ~ "NoHostAvailable". We'll translate it to a `request_execution_exception`.
+          cdc_log.debug("Preimage: translating a (read) `unavailable_exception` to `request_execution_exception` - ", e);
+          throw exceptions::request_execution_exception(exceptions::exception_code::CDC_PREIMAGE_READ_FAILURE,
+                  format("CDC could not achieve the CL of {}. Required {} responses, got {}.",
+                          e.consistency, e.required, e.alive));
+      }
+    }
+
+    static db::consistency_level adjust_cl(db::consistency_level write_cl) {
```

Comment:
Here's a new proposal:

- (ANY, ONE) -> ONE, just because we cannot go lower (but see my remark at the end). The client wants flakily better consistency (TWO, THREE)? OK, boost the preimage in the same way.
- Hitting all other cases with one hammer needs trade-offs. The only question is how much clients expect from the preimage (I don't know the answer). I trust you that QUORUM is the Goodenough's Solution.

Any problems with this solution? For sure, when NO replicas are present, write with CL=ANY can succeed by being captured by hinted handoff. Preimage query will fail - maybe it should insert NULLs instead?

Reply:
I think (ANY, ONE) -> ONE and others -> QUORUM is good enough for now.

Reply:
As we discussed, I force-pushed a yet another solution:
- ANY -> ONE
- ALL -> QUORUM
- (LOCAL_)SERIAL -> (LOCAL_)QUORUM
- for all other cases: WRITE_CL -> PREIMAGE_CL